python/katana/lonestar/analytics/assortativity.py

```diff
@@ -0,0 +1,106 @@
+        steal=True,
+    )
+    return (sum_source_degrees.reduce() / num_edges, sum_destination_degrees.reduce() / num_edges)
+
+
+@do_all_operator()
+def degree_assortativity_coefficient_operator(
+    graph: PropertyGraph,
+    source_degree,
+    source_average,
+    destination_degree,
+    destination_average,
+    product_of_dev: GAccumulator[float],
+    square_of_source_dev: GAccumulator[float],
+    square_of_destination_dev: GAccumulator[float],
+    nid,
+):
+    # deviation of source node from average
+    source_dev = source_degree[nid] - source_average
+    for edge in graph.edges(nid):
+        dst = graph.get_edge_dst(edge)
+        destination_dev = destination_degree[dst] - destination_average
+        product_of_dev.update(source_dev * destination_dev)
+        square_of_source_dev.update(source_dev * source_dev)
+        square_of_destination_dev.update(destination_dev * destination_dev)
+
+
+def degree_assortativity_coefficient(graph: PropertyGraph, x="out", y="in", weight=None):
+    """
+    Calculates and returns the degree assortativity of a given graph.
+    Paramaters:
+       * graph: the PropertyGraph to be analyzed
+       * x: description of degree type to consider for the source node on an edge
+            expected values are "in" or "out"
+       * y: description the degree type to consider for the destination node on an edge
+            expected values are "in" or "out"
+       * weight (optional): edge property to use if using weighted degrees
+    """
+    # get the tables associated with the degree types of the source and destination nodes
+    calculate_degree(graph, "temp_in", "temp_out", weight)
+    source_degree = graph.get_node_property("temp_" + x)
+    destination_degree = graph.get_node_property("temp_" + y)
+
+    # Calculate the average in and out degrees of graph
+    # (with respect to number of edges, not number of nodes)
+    num_edges = graph.num_edges()
+    average = average_degree(graph, num_edges, source_degree, destination_degree)
+    source_average = average[0]
+    destination_average = average[1]
+
+    # Calculate the numerator (product of deviation from mean)
+    # and the factors of the denominator (square deviation from mean)
+    product_of_dev = GAccumulator[float](0)
+    square_of_source_dev = GAccumulator[float](0)
+    square_of_destination_dev = GAccumulator[float](0)
+    do_all(
+        range(graph.num_nodes()),
+        degree_assortativity_coefficient_operator(
+            graph,
+            source_degree,
+            source_average,
+            destination_degree,
+            destination_average,
+            product_of_dev,
+            square_of_source_dev,
+            square_of_destination_dev,
+        ),
+        steal=True,
+        loop_name="degree assortativity coefficient calculation",
+    )
+    graph.remove_node_property("temp_in")
+    graph.remove_node_property("temp_out")
```

Comment:
You should probably put this in a `finally:` block so it runs even if the code above crashes. The `try:` should be up at line 77 I think.